lib/kernel.js

```diff
@@ -66,64 +121,124 @@ export default class Kernel {
   }
 
   addWatchCallback(watchCallback: Function) {
-    this.watchCallbacks.push(watchCallback);
-  }
-
-  _callWatchCallbacks() {
-    this.watchCallbacks.forEach(watchCallback => watchCallback());
+    this.kernel.addWatchCallback(watchCallback);
   }
 
   interrupt() {
-    throw new Error("Kernel: interrupt method not implemented");
+    this.middlewareStack.interrupt();
   }
 
   shutdown() {
-    throw new Error("Kernel: shutdown method not implemented");
+    this.middlewareStack.shutdown();
   }
 
   restart(onRestarted: ?Function) {
-    throw new Error("Kernel: restart method not implemented");
+    this.middlewareStack.restart(onRestarted);
   }
 
   execute(code: string, onResults: Function) {
-    throw new Error("Kernel: execute method not implemented");
+    this.middlewareStack.execute(
+      code,
+      true,
+      this._translateExecuteResults(onResults)
+    );
   }
 
   executeWatch(code: string, onResults: Function) {
-    throw new Error("Kernel: executeWatch method not implemented");
+    this.middlewareStack.execute(
+      code,
+      false,
+      this._translateExecuteResults(onResults)
+    );
+  }
+
+  _translateExecuteResults(onResults: Function) {
```

Comment:
`_parseExecutionResults`?

Reply:
In fact, this method doens't belong to Kernel(=HKernel), this code was(and should be part) of KernelTransport(=Kernel).

Reply:
This method converts from `onResults: (message: JupyterMessageObject, channel: string) => void`, to `onResults: (message: HydrogenInternalUndocumentedFormatObject) => void`. But plugins must never see `HydrogenInternalUndocumentedFormatObject`s. So the conversion must happen **after** all middleware has run (i.e. in the same class as the middleware)

Reply:
OK, now I see the need for changes in `ZMQKernel`, `WSKernel` and `KernelTransport(=Kernel)`.

Let me review again the changes to `ZMQKernel`, `WSKernel` and `KernelTransport(=Kernel)` with that in mind.